File: web/src/services/apiClient.ts
import axios, { type AxiosInstance, type InternalAxiosRequestConfig } from 'axios';

const BASE_URL = import.meta.env.VITE_API_URL ?? '/api';

export const api: AxiosInstance = axios.create({
  baseURL: BASE_URL,
  headers: { 'Content-Type': 'application/json' },
  timeout: 20000,
});

api.interceptors.request.use((config: InternalAxiosRequestConfig) => {
  const token = localStorage.getItem('admin_access_token');
  if (token) config.headers.Authorization = `Bearer ${token}`;
  return config;
});

api.interceptors.response.use(
  (res) => res,
  async (error) => {
    const original = error.config as InternalAxiosRequestConfig & { _retry?: boolean };
    if (error.response?.status === 401 && !original._retry) {
      original._retry = true;
      try {
        const refresh = localStorage.getItem('admin_refresh_token');
        if (!refresh) throw new Error('no refresh token');
        const { data } = await axios.post(`${BASE_URL}/auth/refresh`, { refreshToken: refresh });
        const { accessToken, refreshToken: newRefresh } = data.data as {
          accessToken: string;
          refreshToken: string;
        };
        localStorage.setItem('admin_access_token', accessToken);
        localStorage.setItem('admin_refresh_token', newRefresh);
        original.headers.Authorization = `Bearer ${accessToken}`;
        return api(original);
      } catch {
        localStorage.removeItem('admin_access_token');
        localStorage.removeItem('admin_refresh_token');
        window.location.href = '/login';
      }
    }
    return Promise.reject(error as unknown);
  },
);

export function getApiError(error: unknown): string {
  if (axios.isAxiosError(error)) {
    const data = error.response?.data as { error?: { message?: string } } | undefined;
    return data?.error?.message ?? error.message;
  }
  return 'Unknown error';
}

File: MobileAppFlutter/mobileResponder/lib/core/networking/api_client.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../config/app_config.dart';
import '../storage/secure_storage.dart';
import 'api_exception.dart';

final dioProvider = Provider<Dio>((ref) {
  final storage = ref.read(secureStorageProvider);
  final dio = Dio(BaseOptions(
    baseUrl: AppConfig.apiBaseUrl,
    connectTimeout: const Duration(seconds: 15),
    receiveTimeout: const Duration(seconds: 30),
    headers: {'Content-Type': 'application/json'},
  ));

  dio.interceptors.add(InterceptorsWrapper(
    onRequest: (options, handler) async {
      final token = await storage.getAccessToken();
      if (token != null) options.headers['Authorization'] = 'Bearer $token';
      handler.next(options);
    },
    onError: (error, handler) async {
      if (error.response?.statusCode == 401) {
        try {
          final refreshToken = await storage.getRefreshToken();
          if (refreshToken == null) { await storage.clearTokens(); return handler.next(error); }
          final response = await Dio().post('${AppConfig.apiBaseUrl}/auth/refresh', data: {'refreshToken': refreshToken});
          final data = response.data as Map<String, dynamic>;
          final newAccess = data['data']['accessToken'] as String;
          final newRefresh = data['data']['refreshToken'] as String;
          await storage.saveTokens(accessToken: newAccess, refreshToken: newRefresh);
          error.requestOptions.headers['Authorization'] = 'Bearer $newAccess';
          final retried = await dio.fetch<dynamic>(error.requestOptions);
          return handler.resolve(retried);
        } catch (_) { await storage.clearTokens(); }
      }
      handler.next(error);
    },
  ));
  return dio;
});

ApiException dioErrorToApiException(DioException e) {
  final data = e.response?.data;
  if (data is Map<String, dynamic>) return ApiException.fromMap(data, e.response?.statusCode);
  return ApiException(message: e.message ?? 'Network error', statusCode: e.response?.statusCode);
}
